server/src/main/java/org/apache/druid/server/coordinator/duty/CompactionSegmentIterator.java

```diff
@@ -19,22 +19,45 @@
 
 package org.apache.druid.server.coordinator.duty;
 
-import it.unimi.dsi.fastutil.objects.Object2LongOpenHashMap;
+import org.apache.druid.server.coordinator.CompactionStatistics;
 import org.apache.druid.timeline.DataSegment;
 
 import java.util.Iterator;
 import java.util.List;
+import java.util.Map;
 
 /**
  * Segments in the lists which are the elements of this iterator are sorted according to the natural segment order
  * (see {@link DataSegment#compareTo}).
  */
 public interface CompactionSegmentIterator extends Iterator<List<DataSegment>>
```

Comment:
Thinking about new methods in this interface, I'm not sure if they are good since now every implementation of this interface should track of remaining and processed segments even though the tracking logic will be likely duplicate (even though we have only one implementation yet :slightly_smiling_face:). How about adding `next(Map<String, CompactionStatistics> stats)` so that `CompactSegments` can pass in an appropriate map? Then, it can just iterate over all remaining entries in the iterator without introducing any methods such as `flushAllSegments()` which seems to have a complicated contract. `CompactionSegmentIterator` will not extend `Iterator` anymore in this case.

Reply:
> > > * Why does the response include only one task ID? What will happen if auto compaction issues multiple compaction tasks?
> > 
> > 
> > Hmm... I am thinking of having the UI component show the status (fail, success) of the latest task which could indicate if user action is requried or not.
> > Another idea may be to return a list of all tasks issued during this run (and empty list if no slot). The UI can select which task it want to show. The UI can choose the last task ID or it can calcualte success vs fail % rate of all tasks in last run.
> 
> Hmm, I'm not sure what would be a use case of such information of the tasks issued in each run. Would it be more useful to know how many compaction task failures there have been recently (over a couple of runs)?

I just realized that we not have a different prefix to task name when they are issued manually vs issued automatically by the coordinator (auto compaction). In this case, returning the last task scheduled by auto compaction is not necessary as we can use the existing task api to get the latest task that has the `coordinator-issue` prefix in the name. In fact, we can sort by time and get a history of all the tasks issued by the auto compaction. Thus, the existing task API can already answer question like how many compaction task failures there have been recently (over some time period like hours, days, etc.). 

The information that currently cannot be retrieved is a list for all the tasks issue in the same / latest coordinator run. That provides information you cannot get currently with the ingestion tab since we have no idea which tasks issued in the same run. This allows us to know things like % of tasks succeeded vs. failed vs. canceled in the last run. Moreover, we can then find out  % of tasks succeeded vs. failed vs. canceled in the last run vs previous run (or any other runs). This allows user to see if changes they made (change to configs, specs, etc) between the runs fix any previous issues they were seeing. 
For example...
- run 1 have 100% failed and 0 % success, then user go fix something and now run 2 have 100% success and 0% failed. They can see that the fix they made solve the problem. 
-  run 1 have 100% failed and 0 % success, then user go fix something and now run 2 have 0% success and 100% failed. They can see that the fix did not solve the problem.

Reply:
I don't understand the proposed alternative. What would the `next(Map<String, CompactionStatistics> stats)` do? What map is the `appropriate map` that `CompactSegments` will pass to this method?

Reply:
we can only know the skipped and compacted segments when we are iterating the segment timeline. This code path isn’t called by the next(). It is computed and stored as a QueueEntry into the queue. next() simply just poll the QueueEntry from the queue. This means that the skipped and compacted stats must be in the QueueEntry when the next() call polls the queue. We also do not store an entry into the queue if the List<DataSegment> is empty (no segments to compact). This does not fit well with having to try include the skipped and compacted stats into the QueueEntry. For example, we still need to store and aggregates the skipped and compacted stats when the actual return segment is an empty List (no more segments for the datasource).

tbh I think it is best to keep it this way. Also, since this (the current way or modifying the CompactionSegmentIterator) does not change the API, we can come back and modify it later. Also since NewestSegmentFirstIterator is the only CompactionSegmentIterator right now, the cost for changing isn't too high. Maybe a good time to re-visit this is if we are going to add a new iterator and see what will work best when we have multiple implementations of CompactionSegmentIterator

Reply:
Actually, I have another idea which I really like and I think works well + clean + simple (I hope you like it too).

Removed flushAllSegments() and totalRemainingStatistics()  from the CompactionSegmentIterator. Keep totalCompactedStatistics() and totalSkippedStatistics() in the CompactionSegmentIterator . The stats for totalCompactedStatistics() and totalSkippedStatistics()  is maintain and stored in the CompactionSegmentIterator  (NewestSegmentFirstIterator). I think this makes perfect sense since the already compacted and skipped segments are never returned from the iterator’s next() (in fact, it’s never even put into the queue that backs the iterator). Hence, as we iterate through the already compacted and skipped segments, we should just aggregates the stats within the CompactionSegmentIterator . The totalCompactedStatistics() and totalSkippedStatistics()  will simply just returns the aggregates up to the current point of the iterator being iterated. We will then simply just iterate the iterator as before when creating the compaction task. The List<DataSegment> that is returned is then aggregates into a map in CompactSegments. Similarly, we will continue to iterate the iterator in makeStats after we run out of task slot, then the List<DataSegment> that is returned is then aggregates into a different map in CompactSegments
There are Stats map maintained in both CompactSegments  and CompactionSegmentIterator . But I think it’s much clearer and we separate the “skipped/already compacted” from the “segments returned by the iterator”. Also, no complex contracts in the CompactionSegmentIterator . Login in CompactionSegmentIterator  for totalCompactedStatistics() and totalSkippedStatistics()  is also very simple.